src/Features/CSharp/Portable/BraceCompletion/LessAndGreaterThanBraceCompletionService.cs

```diff
@@ -2,51 +2,55 @@
-using Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion.Sessions;
-using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Shared.Extensions;
-using Microsoft.CodeAnalysis.Text;
-using Microsoft.VisualStudio.Text;
-using Microsoft.VisualStudio.Text.BraceCompletion;
-using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Editor.CSharp.AutomaticCompletion.Sessions
+namespace Microsoft.CodeAnalysis.CSharp.BraceCompletion
 {
-    internal class LessAndGreaterThanCompletionSession : AbstractTokenBraceCompletionSession
+    [Export(LanguageNames.CSharp, typeof(IBraceCompletionService)), Shared]
+    internal class LessAndGreaterThanBraceCompletionService : AbstractBraceCompletionService
     {
-        public LessAndGreaterThanCompletionSession(ISyntaxFactsService syntaxFactsService)
-            : base(syntaxFactsService, (int)SyntaxKind.LessThanToken, (int)SyntaxKind.GreaterThanToken)
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public LessAndGreaterThanBraceCompletionService()
         {
         }
 
-        public override bool CheckOpeningPoint(IBraceCompletionSession session, CancellationToken cancellationToken)
-        {
-            var snapshot = session.SubjectBuffer.CurrentSnapshot;
-            var position = session.OpeningPoint.GetPosition(snapshot);
-            var token = snapshot.FindToken(position, cancellationToken);
+        protected override char OpeningBrace => LessAndGreaterThan.OpenCharacter;
+
+        protected override char ClosingBrace => LessAndGreaterThan.CloseCharacter;
+
+        public override Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+            => CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken);
+
+        protected override bool IsValidOpeningBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.LessThanToken);
+
+        protected override bool IsValidClosingBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.GreaterThanToken);
 
+        protected override async Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
             // check what parser thinks about the newly typed "<" and only proceed if parser thinks it is "<" of 
             // type argument or parameter list
             if (!token.CheckParent<TypeParameterListSyntax>(n => n.LessThanToken == token) &&
                 !token.CheckParent<TypeArgumentListSyntax>(n => n.LessThanToken == token) &&
                 !token.CheckParent<FunctionPointerParameterListSyntax>(n => n.LessThanToken == token) &&
-                !PossibleTypeArgument(snapshot, token, cancellationToken))
+                !await PossibleTypeArgumentAsync(document, token, cancellationToken).ConfigureAwait(false))
```

Comment:
this is strange.  only if *all* of these are not true, we say it's not valid?

Reply:
Not sure on the question here, the check makes sense to me I think.  If the < token is not for any kind of type argument it is not valid for brace completion (since we don't want to complete a < conditional).